aiosupervisor/models/addons.py

```diff
@@ -0,0 +1,342 @@
+
+
+class AppArmor(StrEnum):
+    """AppArmor type."""
+
+    DEFAULT = "default"
+    DISABLE = "disable"
+    PROFILE = "profile"
+
+
+class SupervisorRole(StrEnum):
+    """SupervisorRole type.
+
+    Supervisor may add new Roles in future as a non-breaking change.
+    Those should be returned as string in response models until next update.
+    """
+
+    ADMIN = "admin"
+    BACKUP = "backup"
+    DEFAULT = "default"
+    HOMEASSISTANT = "homeassistant"
+    MANAGER = "manager"
+
+
+class AddonState(StrEnum):
+    """AddonState type."""
+
+    STARTUP = "startup"
+    STARTED = "started"
+    STOPPED = "stopped"
+    UNKNOWN = "unknown"
+    ERROR = "error"
+
+
+# --- OBJECTS ----
+
+
+@dataclass(frozen=True)
+class AddonInfoBaseFields(ABC):
+    """AddonInfoBaseFields ABC type."""
+
+    advanced: bool
+    available: bool
+    build: bool
+    description: str
+    homeassistant: str | None
+    icon: bool
+    logo: bool
+    name: str
+    repository: str
+    slug: str
+    stage: AddonStage | str
+    update_available: bool
+    url: str | None
+    version_latest: str
+    version: str | None
+
+
+@dataclass(frozen=True)
+class AddonInfoStoreBaseFields(ABC):
+    """AddonInfoAllStoreFields ABC type."""
+
+    arch: list[CpuArch]
+    documentation: bool
+
+
+@dataclass(frozen=True)
+class AddonInfoStoreExtFields(ABC):
+    """AddonInfoStoreExtFields ABC type."""
+
+    apparmor: AppArmor
+    auth_api: bool
+    docker_api: bool
+    full_access: bool
+    homeassistant_api: bool
+    host_network: bool
+    host_pid: bool
+    ingress: bool
+    long_description: str | None
+    rating: int
+    signed: bool
+
+    # Hassio is deprecated name for supervisor
+    supervisor_api: bool = field(metadata=field_options(alias="hassio_api"))
+    supervisor_role: SupervisorRole | str = field(
+        metadata=field_options(alias="hassio_role")
+    )
+
+
+@dataclass(frozen=True)
+class AddonInfoStoreExtInstalledBaseFields(ABC):
+    """AddonInfoStoreExtInstalledBaseFields ABC type."""
+
+    detached: bool
+
+
+@dataclass(frozen=True, slots=True)
+class StoreAddon(AddonInfoBaseFields, AddonInfoStoreBaseFields, ResponseData):
+    """StoreAddon type."""
+
+    installed: bool
+
+
+@dataclass(frozen=True, slots=True)
+class StoreAddonComplete(
+    AddonInfoBaseFields,
+    AddonInfoStoreBaseFields,
+    AddonInfoStoreExtFields,
+    AddonInfoStoreExtInstalledBaseFields,
+    ResponseData,
+):
+    """StoreAddonComplete type."""
+
+    installed: bool
+
+
+@dataclass(frozen=True, slots=True)
+class InstalledAddon(
+    AddonInfoBaseFields, AddonInfoStoreExtInstalledBaseFields, ResponseData
+):
+    """InstalledAddon type."""
+
+    state: AddonState
+
+
+@dataclass(frozen=True, slots=True)
+class InstalledAddonComplete(
+    AddonInfoBaseFields,
+    AddonInfoStoreExtInstalledBaseFields,
+    AddonInfoStoreBaseFields,
+    AddonInfoStoreExtFields,
+    ResponseData,
+):
+    """InstalledAddonComplete model."""
+
+    state: AddonState
```

Comment:
So this restricts us to the currently existing add-on state. We did add `startup` relatively recently with [Supervisor PR #4376](https://github.com/home-assistant/supervisor/pull/4376/files#diff-dc1d068b16705906f72e7f53c871400e85e8695d332bde81fb2cf01eec62cf7e). 

IMHO, introducing new states is very on the edge for a "backwards compatible" change :sweat_smile: . So I prefer to not keep this open.

BUT: Looking at the states, it seems the only one which might be missing is `stopping`. Should we maybe just introduce that one and then keep them fixed? :thinking:

Reply:
I guess my thinking was that existing clients (frontend and core mostly) already closely watch and depend on this field. So changing the values here already felt like a breaking change.

I don't know though. I don't really have a good answer here. Supervisor auto-updates so I'm pretty tempted to just put `| str` on all of them just in case, I just don't like the model that makes.

Reply:
I guess I'll just note on here that regardless of whether we change this value in the future or not this one seems like one that it should actually be breaking to add a value. Clients can and will expect to know the state of addons. If we add a new value in the future then they will be clueless as to what state it is in.

Out of all the enums the ones for State of Supervisor and Addons seem like the most obvious candidate for even additive changes are breaking. The others I'm less sure about.